{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Virus UR clusters logos\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Computing the clusters\n",
    "\n",
    "The code below supports two clustering tools:\n",
    "1. Starcode. This has three different modes (clustering algorithms).\n",
    "2. MeshClust.\n",
    "\n",
    "Set the parameters in the cell below. Then execute, and then execute the third cell to generate the logos figures, that are also captured in files. The second cell can be used to display the clusters to stdout but is not really needed.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os, sys\n",
    "import re\n",
    "from Bio import motifs, SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import matplotlib.pyplot as plt\n",
    "from collections import OrderedDict\n",
    "import subprocess\n",
    "from pprint import pprint\n",
    "\n",
    "import mysequtils as myut\n",
    "import myseqclusters as mycl\n",
    "import myseq_logo as mysl\n",
    "import pald_funcs as mypal # my palindrome functions\n",
    "\n",
    "\n",
    "def run_clst_tool(clst_tool, infile, outfile, exe_command, flags):\n",
    "    '''This function calls the corresponding clustering tool'''\n",
    "    command = exe_command+flags+' -i ' + infile  + ' -o ' + outfile if clst_tool=='strc' \\\n",
    "    else ' '.join([exe_command, infile, flags, '--output '+outfile])\n",
    "    return subprocess.getstatusoutput(command)   \n",
    "\n",
    "def parse_clst_tool_output(clst_tool, outfile):\n",
    "    '''This function parses the output file generated by the \n",
    "    cluster tool.'''\n",
    "    return mycl.parse_starcode_cls_out(outfile) if clst_tool=='strc' else mycl.parse_meshclsut_cls_out(outfile)\n",
    "\n",
    "# ===================================================================================================\n",
    "\n",
    "mlens = [3, 4, 5]\n",
    "num_tst_vtaxids = 10 # the number of vtaxids with the most common UR to process\n",
    "\n",
    "# Clustering tool\n",
    "clst_tool = 'strc' # select a cluster tool: 'strc' for Starcode or 'mesh' for MeshClust\n",
    "\n",
    "# starcode parameters\n",
    "str_exe = '~/work/mystuff/tools/starcode/starcode'\n",
    "str_clst_type = 'sp' # 'mp' for message-passing algorithm, 'sp' for sphere or 'cc' for connected-components\n",
    "# MeshClust parameters\n",
    "mesh_exe = '~/work/mystuff/tools/MeShClust-master/src/cluster/meshclust'\n",
    "mesh_id = 0.5\n",
    "\n",
    "clst_bfname = 'ur' # file name is clst_bfname_<vtaxid>_<mlen>.seq for UR and .clst for clusters\n",
    "clst_path = './clst_data' # will be appended with clst_tool, and in case \n",
    "                          # of Starcode, with _mp or _ph based on str_clst_type\n",
    "# UR information\n",
    "base_path = '/Users/yoramzarai/work/school/Simulation/Viruses/VUR/viruses/'\n",
    "dtype_rnd = {'vtaxid':int, 'num_cds':int, 'signf_common':str, 'signf_F1':str, 'signf_F2':str, 'signf_F3':str}\n",
    "backg = {'A':0.25,'C':0.25,'G':0.25,'T':0.25}  # background for PSSM \n",
    "base_save_file = '' # 'v_most_URs_diff' code appends '.png' to save the figure and '.txt' to save the prints. \n",
    "                    #  Set to '' to disable \n",
    "# =============================================================================================================\n",
    "\n",
    "if clst_tool=='strc':\n",
    "    print('Clustering tool selected is Starcode')\n",
    "    print('Clustering using {} algorithm'.format(str_clst_type))\n",
    "    clst_path += '_'+clst_tool+'_'+str_clst_type \n",
    "    clst_flags = ' -s' if str_clst_type=='sp' else ' -r 1' if str_clst_type=='mp' else ' -c -d 1'\n",
    "    clst_flags += ' --print-clusters'\n",
    "    clst_exe = str_exe\n",
    "elif clst_tool=='mesh':\n",
    "    print('Clustering tool selected is MeshClust')\n",
    "    clst_path += '_'+clst_tool\n",
    "    clst_flags = ' --id '+str(mesh_id)\n",
    "    clst_exe = mesh_exe\n",
    "else: \n",
    "    print('Cluster tool {} not supported ! Exiting...'.format(clst_tool))\n",
    "    sys.exit()\n",
    "\n",
    "os.makedirs(clst_path, exist_ok=True)\n",
    "mmlens = ['m'+str(m) for m in mlens]\n",
    "# handle for text file or sys.stdout\n",
    "ftxt = open(base_save_file+'.txt', 'wt') if base_save_file!='' else sys.stdout\n",
    "bfdata = os.path.join(clst_path, clst_bfname)\n",
    "\n",
    "'''keys are (vtaxid, mlen), and values are dictionaries containing\n",
    "clusters information (where keys are the centroids and values are the corresponding\n",
    "cluster's sequences)'''\n",
    "ur_clst_info = OrderedDict()\n",
    "for j, mlen in enumerate(mlens):\n",
    "    rnd_model = 'dnt_samp' if mlen==3 else 'syn_perm+dnt_samp'\n",
    "    mmlen = mmlens[j]\n",
    "    xlsx_file = os.path.join(base_path, mmlens[j], rnd_model, 'Results', 'vsignf'+'_'+rnd_model+'_'+mmlens[j]+'.xlsx')\n",
    "    df_vrnd = pd.read_excel(xlsx_file, header=0, dtype=dtype_rnd)\n",
    "    vtaxids = df_vrnd.loc[:, 'vtaxid']  # vtaxid Series\n",
    "    # find the vtaxids with the most common URs\n",
    "    sgf = df_vrnd.loc[:,'signf_common']  # all common URs per vtaxid\n",
    "    tmp = sgf.apply(lambda x:len(x.split('|'))) # this works on the values of sgf (which are the URs)\n",
    "    srt_tmp = tmp.sort_values(ascending=False)\n",
    "    for i in range(num_tst_vtaxids):\n",
    "        indx = srt_tmp.index[i]  # vtaxid index in srt_tmp\n",
    "        vtaxid = vtaxids.iloc[indx]\n",
    "        ur = df_vrnd.iloc[indx]['signf_common'].split('|')\n",
    "        eql_nt = {s for s in ur if len(set(s))==1}  # URs with the same nucleotides (e.g., AAAA)\n",
    "        gc_nt = {s for s in ur if re.search('[CG]{'+str(mlen)+'}',s)}  # URs that contain only G and/or C\n",
    "        at_nt = {s for s in ur if re.search('[AT]{'+str(mlen)+'}',s)}  # URs that contain only A and/or T\n",
    "        \n",
    "        # CHOOSE here what is proc_nt (the URs to show logos)\n",
    "        #proc_nt = set(ur)-eql_nt # all UR other than the ones with equal NT\n",
    "        proc_nt = set(ur)\n",
    "        #proc_nt = gc_nt\n",
    "        #proc_nt = at_nt\n",
    "        #print('{}:\\n\\teql={} \\n\\tgc={}\\n\\tat={}\\n\\tproc={}'.format(vtaxid, eql_nt, gc_nt, at_nt, proc_nt), file=ftxt)\n",
    "        if proc_nt and list(proc_nt)[0]!='nan':  # 'nan' is generated by pd.read_excel for empty cells\n",
    "            _, _, _, m, _ = mysl.compute_pssm(proc_nt, backg, backg)\n",
    "            rel_info = mysl.calc_rel_info(m, 'no')\n",
    "            \n",
    "            # clustering\n",
    "            infile = '_'.join([bfdata, str(vtaxid), str(mlen)]) + '.seq'\n",
    "            outfile = re.sub('.seq', '.clst', infile)\n",
    "            #myut.seq2plain_text_file(proc_nt, infile)  # generate input file\n",
    "            myut.create_fasta_file(infile, proc_nt, proc_nt)  # generate input file\n",
    "            # run the cluster tool\n",
    "            ret, sout = run_clst_tool(clst_tool, infile, outfile, clst_exe, clst_flags)\n",
    "            #print(sout)\n",
    "            if ret == 0:\n",
    "                # parse starcode output\n",
    "                ur_clst_info[(vtaxid, mlen)] = parse_clst_tool_output(clst_tool, outfile)\n",
    "            else:\n",
    "                print('Clustering tool {} errored and returned: {}'.format(clst_tool, sout))\n",
    "            "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Displaying sequence logos of clusters data \n",
    "\n",
    "NOTE: This cell doesn't have to be executed. It is only used to print the clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# MAKE SURE THE CELL ABOVE RUNS FIRST !!!\n",
    "\n",
    "# ur_clst_info contains clusters information\n",
    "for k in ur_clst_info:\n",
    "    print('clusters for {}:'.format(k))\n",
    "    val = ur_clst_info[k]\n",
    "    print('number of subplots = ', sum( [len(val[kk])>1 for kk in val.keys()]))\n",
    "    for i, kk in enumerate(val.keys()):\n",
    "        print('  C{} ({} seq): {}'.format(i,len(val[kk]), val[kk]))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Displaying the clusters' logos\n",
    "\n",
    "MAKE SURE THE FIRST CELL RUNS FIRST !!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import myseq_logo as mysl\n",
    "import pald_funcs as mypal # my palindrome functions\n",
    "from Bio import motifs, SeqIO\n",
    "from Bio.Seq import Seq\n",
    "\n",
    "figs_path = './clst_figs'  # will be appended with _mp or _ph based on str_clst_type\n",
    "figs_bfname = 'ur' # file name is figs_bfname_<vtaxid>_<mlen>.png\n",
    "backg = {'A':0.25,'C':0.25,'G':0.25,'T':0.25}  # background for PSSM \n",
    "# =========================================================================================\n",
    "figs_path += '_'+clst_tool\n",
    "if clst_tool=='strc': figs_path += '_'+str_clst_type\n",
    "\n",
    "os.makedirs(figs_path, exist_ok=True)\n",
    "base_fname = os.path.join(figs_path, figs_bfname)\n",
    "\n",
    "# ur_clst_info contains clusters information\n",
    "for k in ur_clst_info:\n",
    "    vtaxid, mlen = k\n",
    "    fname = '_'.join([base_fname, str(vtaxid), str(mlen)]) + '.png'\n",
    "    num_urs = sum([len(v) for v in ur_clst_info[k].values()])\n",
    "    # processing clusters that contain at least two sequences\n",
    "    clsts = [v for v in ur_clst_info[k].values() if len(v)>1]\n",
    "    fig, axs = plt.subplots(len(clsts), 1, figsize=(6, 3*len(clsts)), squeeze=False, constrained_layout=True)\n",
    "    fig.suptitle('m'+str(mlen)+':'+str(vtaxid)+' ('+str(num_urs)+' URs)', fontsize=25, color='blue')\n",
    "    for i, c in enumerate(clsts):\n",
    "        _, _, _, m, _ = mysl.compute_pssm(c, backg, backg)\n",
    "        rel_info = mysl.calc_rel_info(m, 'no')\n",
    "        maxy = mysl.gen_nt_sequence_logo(axs[i,0], rel_info)\n",
    "        #axs[i,0].set_title('m'+str(mlen)+' : '+str(vtaxid) + \\\n",
    "        axs[i,0].set_title('('+str(len(c))+' cURs)',\\\n",
    "                               fontsize=15, color='black')\n",
    "        axs[i,0].set_ylabel('Bits')\n",
    "        #axs[i,j].set_xlabel('NT index')\n",
    "        axs[i,0].axis([0.5, mlen+0.5, 0, maxy])\n",
    "        axs[i,0].set_xticks(range(1, mlen+1))\n",
    "        \n",
    "    #plt.tight_layout()\n",
    "    #plt.subplots_adjust(wspace=0.8, top=0.8)\n",
    "    # save plot\n",
    "    plt.savefig(fname, dpi=200)\n",
    "    path = !pwd  # shell command\n",
    "    print('Figure saved in {}/{}'.format(path[0], fname))\n",
    "\n",
    "\n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Scratch Pad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SCRATCH PAD\n",
    "########################\n",
    "from pprint import pprint\n",
    "\n",
    "a = (0,1)\n",
    "b, c = a\n",
    "print(b,c)\n",
    "\n",
    "\n",
    "d = { 'a': [1,2], 'b': [1,2,3], 'c': [2], 'd': [1,2,3, 4, 5]}\n",
    "pprint(d)\n",
    "\n",
    "onlyv = [v for k,v in d.items() if len(v)>1]\n",
    "for i, c in enumerate(onlyv):\n",
    "    print(i, c)\n",
    "\n",
    "k = {k: v for k,v in d.items() if len(v)>1}\n",
    "pprint(k)\n",
    "print(len(k))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
